### DISCUSSION TITLE
Is Hyper the best way to use solid.js from langauges that doesn't support jsx?

### DISCUSSION THREAD

**Discussion by user753** [solidjs/solid#774] [Q&A] [answered]
For example if I want to use solid.js with languages like kotlin, dart, f#. Would Hyper DOM Expressions be the best way to do it?
Is there any way to somehow still get the advantages from jsx compiler optimizations?

**Comment by atk** [comment]
The only way to get it would be to replicate the DOM expressions transformations yourself in whatever language you are using.

**Comment by trusktr** [comment, marked as answer]
The ``` html`` ``` template string tag has better performance than the `h()` functions. `html` applies the same compiler optimizations as JSX (just that compilation happens at runtime once for each template (not once per component instantiation, but once per source position and then re-used for everyone component instantiation, thanks to the magic of ECMAScript template tags).

If another language has something like template tags including caches string parts (the magic), then use that.

But if not, then hyperscript-style `h()` functions is the way to go. Those are less optimized because the template that will be constructed is not known yo front, but happens across multiple `h()` over the whole tree that is being created.

It may be possible to cache the tree structure results, and to re-cache conditional branches as they are discovered at runtime, but that sort of thing hasn't been implemented in Solid yet.

**Reply by user753** [reply]
Is there a place where I can read about all these html`` and jxs compiler optimizations and how solid compiler works?

**Reply by trusktr** [reply]
- [``` html`` ```](https://github.com/solidjs/solid/tree/main/packages/solid/html)
- [`h()`](https://github.com/solidjs/solid/tree/main/packages/solid/h)
- [JSX](https://github.com/solidjs/solid/tree/main/packages/babel-preset-solid)

Other than that, the [Solid playground](https://playground.solidjs.com/) will show you the JSX output that you can study.
